### PR TITLE
FX tracing improvement

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

This PR improves significantly the way transformers models are traced by the HFTracer (`torch.fx`).
This has 2 major consequences:

- More model architectures can be supported
- When a model can be traced, the resulting GraphModule can take any input shapes out of the box (compared to what was done before where a lot of work was needed to enable dynamic axes for a given model), this is both easier and less bug prone.

Because of these changes the `symbolic_trace` signature becomes easier: 
`symbolic_trace(model: PreTrainedModel, input_names: Optional[List[str]] = None) -> GraphModule`

There is no need to specify the batch size, the sequence length or the number of choices (for multiple-choice) anymore.
The same thing can be said about the `HFTracer`, which can be instantiated exactly the same way as the regular `torch.fx.Tracer` can.

### CODE REVIEW

**Comment by LysandreJik** [comment]
Hey, thanks for your PR @michaelbenayoun ! It seems there are a few failing tests (1096 :smile:), could you take a look at it?

**Comment by michaelbenayoun** [comment]
Currently looking into it!
Sorry about that.

**Comment by michaelbenayoun** [comment]
Fixed!

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -266,26 +287,29 @@ def get_extended_attention_mask(self, attention_mask: Tensor, input_shape: Tuple
...
+                #         [
+                #             torch.ones(
+                #                 (batch_size, seq_length, prefix_seq_len), device=device, dtype=causal_mask.dtype
+                #             ),
+                #             causal_mask,
+                #         ],
+                #         axis=-1,
+                #     )
+
+                # extended_attention_mask = causal_mask[:, None, :, :] * attention_mask[:, None, None, :]
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:309]: This needs cleaning up.

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -1749,7 +1773,7 @@ def __init__(self, nf, nx):
     def forward(self, x):
         size_out = x.size()[:-1] + (self.nf,)
         x = torch.addmm(self.bias, x.view(-1, x.size(-1)), self.weight)
-        x = x.view(*size_out)
+        x = x.view(size_out)
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:1776]: Is that change absolutely necessary? Are we sure it's 100% backward compatible with previous versions of PyTorch?

File: src/transformers/models/xlnet/modeling_xlnet.py [hunk 3]
```diff
@@ -1443,7 +1443,7 @@ def forward(
             output_attentions=output_attentions,
             output_hidden_states=output_hidden_states,
             return_dict=return_dict,
-            **kwargs,
+            # **kwargs,
```
Comment (sgugger) [inline src/transformers/models/xlnet/modeling_xlnet.py:1446]: Why are those commented out? This will break backward compatibility with the deprecation cycle inside XLNet for `use_cache`.

File: tests/test_modeling_bart.py [hunk 4]
```diff
@@ -411,6 +411,7 @@ class BartModelTest(ModelTesterMixin, GenerationTesterMixin, unittest.TestCase):
         if is_torch_available()
         else ()
     )
+    # fx_ready_model_classes = all_model_classes
```
Comment (sgugger) [inline tests/test_modeling_bart.py:414]: Should be removed?

File: tests/test_modeling_layoutlm.py [hunk 5]
```diff
@@ -215,6 +215,7 @@ class LayoutLMModelTest(ModelTesterMixin, unittest.TestCase):
         if is_torch_available()
         else None
     )
+    # fx_ready_model_classes = all_model_classes
```
Comment (sgugger) [inline tests/test_modeling_layoutlm.py:218]: Should be removed?

**Review by sgugger — COMMENTED** [review]
I'm not too comfortable with some of the changes in the models, especially XLNet, apart from that, the PR looks good.

In the tests, the `fx_ready_model_classes` seems to always be set to `all_model_classes`, so maybe it's time to use a boolean flag instead of a list of classes, if we always test all classes?

File: src/transformers/models/xlnet/modeling_xlnet.py (hunk 3 above)
Comment (michaelbenayoun) [reply src/transformers/models/xlnet/modeling_xlnet.py:1446]: This makes tracing fail, I readded the kwargs, and will add support for XLNet once this is solved.

Comment (michaelbenayoun) [reply src/transformers/models/xlnet/modeling_xlnet.py:1446]: This makes tracing fails, I re-added the kwargs and will enable tracing for XLNet once this is solved.

**Review by thomasw21 — COMMENTED** [review]
Haven't had the time to look in depth. I'll review more when I'll have some more bandwidth

File: src/transformers/modeling_utils.py (hunk 2 above)
Comment (michaelbenayoun) [reply src/transformers/modeling_utils.py:1756]: Good point!
So I checked, the older version for which a documentation is available (0.1.12_2) says for [view](https://pytorch.org/docs/0.1.12/tensors.html?highlight=view#torch.Tensor.view): 

> Parameters:	args (torch.Size or int...) – Desired size

Which seems to be exactly the same for [1.10:](https://pytorch.org/docs/1.10.0/generated/torch.Tensor.view.html?highlight=view#torch.Tensor.view)

> shape (torch.Size or int...) – the desired size

So I think it is okay. The reason for this is that when unpacking args instead of giving a list, it iterates over args, but args is actually a `torch.Proxy` during tracing, and cannot be iterated over easily, causing tracing to fail.

Comment (sgugger) [reply src/transformers/modeling_utils.py:1756]: Thanks for checking!

**Comment by michaelbenayoun** [comment]
Unstale comment

**Comment by michaelbenayoun** [comment]
I am planning to try another approach to make both the code easier, and the tracing process more clean, this will allow to add other models as well as limit the number of bugs.
In the mean time, I think this can be merged as a few issues were posted to have `symbolic_trace` working for Pytorch 1.10.

File: src/transformers/commands/add_new_model_like.py [hunk 6]
```diff
@@ -1189,6 +1189,13 @@ def create_new_model_like(
             if "tokenization" not in str(f) and "processor" not in str(f) and "feature_extraction" not in str(f)
         ]
 
+    def disable_fx_test(filename: Path):
```
Comment (michaelbenayoun) [inline src/transformers/commands/add_new_model_like.py:1192]: What do you think of this @sgugger ?

The reason I added that is because `symbolic_trace` checks the model class before trying to trace the model to make sure it is supported. 
Because the tests are copied, if a new model is created from a supported model for symbolic tracing, the test file will contain something like `fx_ready = True` which will trigger the torch.fx tests, all of them failing because the model class is not in the list of the supported models. 
I do not think this is a good approach to automatically add the new model class to the supported models because the model implementation can be changed, so I thought that disabling the test and printing some message was a better option.

File: src/transformers/commands/add_new_model_like.py [hunk 7]
```diff
@@ -1189,6 +1189,13 @@ def create_new_model_like(
...
+        with open(filename) as fp:
+            content = fp.read()
+        with open(filename, "w") as fp:
+            new_content = re.sub(r"fx_ready\s*=\s*True", "fx_ready = False", content)
```
Comment (sgugger) [inline src/transformers/commands/add_new_model_like.py:1196]: Nit, this line should go before the second `with`.

File: src/transformers/commands/add_new_model_like.py (hunk 6 above)
Comment (sgugger) [reply src/transformers/commands/add_new_model_like.py]: Works for me!

File: src/transformers/commands/add_new_model_like.py (hunk 7 above)
Comment (michaelbenayoun) [reply src/transformers/commands/add_new_model_like.py]: Done!

**Review by LysandreJik — APPROVED** [review]
This looks good to me as long as it's 100% backwards compatible.

Pinging @patrickvonplaten and @patil-suraj for a quick look as it touches to a lot of different models.

File: src/transformers/file_utils.py [hunk 8]
```diff
@@ -322,7 +322,7 @@
 HUGGINGFACE_CO_PREFIX = HUGGINGFACE_CO_RESOLVE_ENDPOINT + "/{model_id}/resolve/{revision}/{filename}"
 
 # This is the version of torch required to run torch.fx features and torch.onnx with dictionary inputs.
-TORCH_FX_REQUIRED_VERSION = version.parse("1.9")
+TORCH_FX_REQUIRED_VERSION = version.parse("1.10")
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply src/transformers/file_utils.py:325]: I can check for torch 1.9, the plan from now on is to support torch 1.10 + as fx became stable starting at this version (still need to validate that with pytorch team).

File: src/transformers/commands/add_new_model_like.py [hunk 9]
```diff
@@ -1201,6 +1211,13 @@ def create_new_model_like(
             dest_file=dest_file,
             add_copied_from=False,
         )
+        disabled_fx_test = disabled_fx_test | disable_fx_test(dest_file)
+
+    if disabled_fx_test:
+        print(
+            "The tests for symbolic tracing with torch.fx were disabled, you can add those once symbolic tracing works "
+            "for your new model."
+        )
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply src/transformers/commands/add_new_model_like.py:1220]: I followed what was done in the script, but can definitely change that to logger if needed.

File: src/transformers/file_utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (LysandreJik) [reply src/transformers/file_utils.py:325]: Sure, sounds good to me

[the remark this answers is not in the document]
Comment (byronyi) [reply src/transformers/file_utils.py:325]: And you probably need to change [this line](https://github.com/huggingface/transformers/blob/90166121ee100838507077c2ac80b69660f25afa/src/transformers/file_utils.py#L408) from `==` to `>=`.

**Review by patil-suraj — COMMENTED** [review]
Went through all the modeling changes and it looks good to me!

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 10]
```diff
@@ -193,7 +193,7 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
         attn_weights = torch.matmul(query, key.transpose(-1, -2))
 
         if self.scale_attn_weights:
-            attn_weights = attn_weights / (float(value.size(-1)) ** 0.5)
+            attn_weights = attn_weights / (value.size(-1) ** 0.5)
```
[the remark this answers is not in the document]
Comment (hyunwoongko) [reply src/transformers/models/gpt2/modeling_gpt2.py:196]: In my opinion, this doesn't cause any problems.

When we do tracing, python values cause several problems. 
I don't think there is any reason to change this value to a Python value.

File: src/transformers/modeling_utils.py [hunk 11]
```diff
@@ -242,6 +242,27 @@ def invert_attention_mask(self, encoder_attention_mask: Tensor) -> Tensor:
 
         return encoder_extended_attention_mask
 
+    def create_extended_attention_mask_for_decoder(self, input_shape, attention_mask, device):
+        batch_size, seq_length = input_shape
+        seq_ids = torch.arange(seq_length, device=device)
+        causal_mask = seq_ids[None, None, :].repeat(batch_size, seq_length, 1) <= seq_ids[None, :, None]
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/modeling_utils.py:253]: Would be nice if we keep the code as is for now to make sure to not break anything here accidentally. Could you also run T5's and Bart's SLOW tests to be sure nothing is broken with the attention mask?

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 12]
```diff
@@ -1410,7 +1410,7 @@ def forward(
                     f"unexpected if using padding tokens in conjunction with `inputs_embeds.`"
                 )
 
-        pooled_logits = logits[range(batch_size), sequence_lengths]
+        pooled_logits = logits[torch.arange(batch_size), sequence_lengths]
```
Comment (patrickvonplaten) [inline src/transformers/models/gpt2/modeling_gpt2.py:1413]:
```suggestion
        pooled_logits = logits[torch.arange(batch_size, device=self.device), sequence_lengths]
```
We need to make sure the tensor is on the same device no?

File: src/transformers/models/gptj/modeling_gptj.py [hunk 13]
```diff
@@ -945,7 +945,7 @@ def forward(
                     f"unexpected if using padding tokens in conjunction with `inputs_embeds.`"
                 )
 
-        pooled_logits = logits[range(batch_size), sequence_lengths]
+        pooled_logits = logits[torch.arange(batch_size), sequence_lengths]
```
Comment (patrickvonplaten) [inline src/transformers/models/gptj/modeling_gptj.py:948]: same here

File: tests/test_modeling_bert.py [hunk 14]
```diff
@@ -444,8 +444,7 @@ class BertModelTest(ModelTesterMixin, GenerationTesterMixin, unittest.TestCase):
         else ()
     )
     all_generative_model_classes = (BertLMHeadModel,) if is_torch_available() else ()
-    fx_ready_model_classes = all_model_classes
-    fx_dynamic_ready_model_classes = all_model_classes
+    fx_ready = True
```
Comment (patrickvonplaten) [inline tests/test_modeling_bert.py:447]: (nit) not a huge fan of the name `fx_ready` - does that mean `fx_compatible`?

**Review by patrickvonplaten — COMMENTED** [review]
Left some comments, but in general this looks good to me as well

File: src/transformers/models/gpt2/modeling_gpt2.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply src/transformers/models/gpt2/modeling_gpt2.py:196]: This change seems to cause the fail on training gpt-2 with ONNX Runtime backend. Link to the reported issue [#11279](https://github.com/microsoft/onnxruntime/issues/11279).
